### PR TITLE
Upgrade to .NET 8

### PR DESCRIPTION
**Pull request by IEvangelist**
I just now saw #186, I didn't realize before that there was an effort in place to do this work. There are parts of that PR that look good—@luisquintanilla how should we proceed?

I pulled in the changes from the PR that I liked. ~Let's add the CPM in a separate PR, since versions weren't bumped~.

I've bumped the versions, and I'm updating this PR to utilize the CPM feature since it does improve the `PacakgeReference` experience.

### CODE REVIEW

**Comment by luisquintanilla** [comment]
Thanks @IEvangelist

File: app/backend/Extensions/SearchClientExtensions.cs [hunk 1]
```diff
@@ -35,21 +35,23 @@ internal static class SearchClientExtensions
...
+        //    var vectorQuery = new SearchQueryVector
+        //    {
+        //        // if semantic ranker is enabled, we need to set the rank to a large number to get more
+        //        // candidates for semantic reranking
+        //        KNearestNeighborsCount = useSemanticRanker ? 50 : top,
+        //        Value = embedding,
+        //    };
+        //    vectorQuery.Fields.Add("embedding");
+        //    searchOption.Vectors.Add(vectorQuery);
+        //}
```
Comment (IEvangelist) [inline app/backend/Extensions/SearchClientExtensions.cs:51]:
```suggestion
        if (embedding != null && overrides?.RetrievalMode != "Text")
        {
            var k = useSemanticRanker ? 50 : top;
            var vectorQuery = new RawVectorQuery
            {
                // if semantic ranker is enabled, we need to set the rank to a large number to get more
                // candidates for semantic reranking
                KNearestNeighborsCount = useSemanticRanker ? 50 : top,
                Value = embedding,
            };
            vectorQuery.Fields.Add("embedding");
            searchOption.Vectors.Add(vectorQuery);
        }
```

**Comment by luisquintanilla** [comment]
> I just now saw #186, I didn't realize before that there was an effort in place to do this work. There are parts of that PR that look good—@luisquintanilla how should we proceed?
> 
> I pulled in the changes from the PR that I liked. ~Let's add the CPM in a separate PR, since versions weren't bumped~.
> 
> I've bumped the versions, and I'm updating this PR to utilize the CPM feature since it does improve the `PacakgeReference` experience.

Sounds good @IEvangelist. Thanks.

Which parts from #186 are not in this PR?

**Comment by aaronpowell** [comment]
Just be aware that #186 builds on top of #184 so there are some unrelated changes are the VS Code experience in the PR.

**Comment by aaronpowell** [comment]
I noticed that this PR also upgrades to C# 12 preview - I'd suggest not doing that in this PR but in a follow-up one to ensure that it's easier to review the .NET 8 specific changes without also having language changes in there.

Small and targeted PRs are easier to review and remediate.

File: app/backend/Extensions/SearchClientExtensions.cs [hunk 2]
```diff
@@ -38,18 +38,19 @@ internal static class SearchClientExtensions
         if (embedding != null && overrides?.RetrievalMode != "Text")
         {
             var k = useSemanticRanker ? 50 : top;
-            var vectorQuery = new SearchQueryVector
+            var vectorQuery = new RawVectorQuery
```
Comment (LittleLittleCloud) [inline app/backend/Extensions/SearchClientExtensions.cs:41]: Oh just see that you made the change, sorry for the late reply

**Comment by luisquintanilla** [comment]
@LittleLittleCloud do these changes affect the App Insight / Authentication changes you've made?

**Comment by luisquintanilla** [comment]
Cool. Let's merge those in first then and do a quick test and follow of these changes.

**Comment by aaronpowell** [comment]
Shall we close #186 in favour of this?

Yesterday I noticed a problem in my PR with VS Code debugging which I addressed in https://github.com/Azure-Samples/azure-search-openai-demo-csharp/pull/186/commits/178aa843c4f8149e779387ff3f3931aa8c2eb492 (although there's another problem that I'm digging through)

**Comment by luisquintanilla** [comment]
> Shall we close #186 in favour of this?
> 
> Yesterday I noticed a problem in my PR with VS Code debugging which I addressed in [178aa84](https://github.com/Azure-Samples/azure-search-openai-demo-csharp/commit/178aa843c4f8149e779387ff3f3931aa8c2eb492) (although there's another problem that I'm digging through)

I'm okay with that. Should #184 be closed as well since #186 built on top of it?

**Comment by aaronpowell** [comment]
> > Shall we close #186 in favour of this?
> > Yesterday I noticed a problem in my PR with VS Code debugging which I addressed in [178aa84](https://github.com/Azure-Samples/azure-search-openai-demo-csharp/commit/178aa843c4f8149e779387ff3f3931aa8c2eb492) (although there's another problem that I'm digging through)
> 
> I'm okay with that. Should #184 be closed as well since #186 built on top of it?

I'd merge #184 first as the changes it introduces are unrelated to the .NET 8 upgrade, so it would give better history tracking of when/why changes were made.

**Review by LittleLittleCloud — APPROVED** [review]

**Review by LittleLittleCloud — APPROVED** [review]

**Pull request merged**